Bezier curve (#283)
javascript-tutorial/it.javascript.info: 7-animation/1-bezier-curve/article.md

# Curve di Bezier

Le curve di Bezier sono utilizzate in computer grafica per disegnare forme, per animazioni CSS ed in molti altri posti.

Sono veramente molto semplici, vale la pena studiarle per poi trovarsi a proprio agio nel mondo della grafica vettoriale e con animazioni avanzate.

## Punti di controllo

Una [curva di Bezier](https://en.wikipedia.org/wiki/B%C3%A9zier_curve) è composta da punti di controllo.

Possono essercene 2, 3, 4 o più.

Ad esempio, una curva a due punti:

![](bezier2.svg)

Una curva a tre punti:

![](bezier3.svg)

Una curva a quattro punti:

![](bezier4.svg)

Guardando da vicino queste curve, possiamo subito notare:

1. **I punti non sono sempre nella curva.** Questo è perfettamente normale, più avanti vedremo come viene costruita la curva.
2. **Il grado della curva equivale al numero di punti meno uno**.
Con due punti avremo una curva lineare (una retta), con tre punti, avremo una curva di grado due (una parabola), con quattro punti, avremo una curva di grado tre.
3. **Una curva è sempre definita internamente all'[inviluppo convesso](https://en.wikipedia.org/wiki/Convex_hull) dei punti di controllo:**

    ![](bezier4-e.svg) ![](bezier3-e.svg)

> Proprio grazie a quest'ultima proprietà, in computer grafica è possibile ottimizzare le verifiche di interesezione. Se due inviluppi convessi non si intersecano, allora neanche le curve lo fanno. Quindi le verifiche sugli inviluppi convessi possono dare degli esiti più rapidi nella verifica della "non intersezione". La verifica di intersezione tra inviluppi convessi è molto più semplice, poiché questi sono rettangoli, triangoli e cosi via (vedi la figura sopra), le quali sono figure molto più semplici rispetto alle curve.

Reviewer: ```suggestion
Proprio grazie a quest'ultima proprietà, in computer grafica è possibile ottimizzare i test di intersezione. Se due inviluppi convessi non si intersecano, allora neanche le curve lo fanno. Quindi le verifiche sugli inviluppi convessi possono dare degli esiti più rapidi nella verifica della "non intersezione". La verifica di intersezione tra inviluppi convessi è molto più semplice, poiché questi sono rettangoli, triangoli e cosi via (vedi la figura sopra), le quali sono figure molto più semplici rispetto alle curve.
```

> **Il principale vantaggio delle curve di Bezier in grafica, è che muovendo i punti le curve cambiano *in maniera piuttosto intuitiva*.**

Reviewer: ```suggestion
**In grafica, ll principale vantaggio delle curve di Bezier è che muovendo i punti, le curve cambiano *in maniera piuttosto intuitiva*.**
```

**Come puoi notare, la curva si estende lungo le linee tangenziali 1 -> 2 e 3 -> 4.**

Dopo aver fatto un po' di pratica diventa ovvio come posizionare i punti per ottenere la curva desiderata. E connettendo diverse curve possiamo ottenere praticamente qualsiasi cosa.

Qui vediamo alcuni esempi:

![](bezier-car.svg) ![](bezier-letter.svg) ![](bezier-vase.svg)

## L'algoritmo di De Casteljau's

Esiste una formula matematica per la definizione delle curve di Bezier, ma la vederemo tra un po', perché [l'algoritmo di De Casteljau's](https://en.wikipedia.org/wiki/De_Casteljau%27s_algorithm) è identico e dimostra a livello visivo come vengono definite.

Come prima cosa, vediamo l'esempio con 3 punti.

Qui vediamo la dimostrazione, seguita dalla spiegazione.

I punti di controllo (1,2 e 3) possono essere mossi utilizzando il mouse. Premete il pulsante "Play" per avviare l'esempio.

[iframe src="demo.svg?p=0,0,0.5,1,1,0&animate=1" height=370]

**L'algoritmo di De Casteljau's per la costruzione di una curva di Bezier a 3 punti:**

1. Traccia i punti di controllo. Nella dimostrazione vista sopra, questi sono etichettati come: `1`, `2`, `3`.

> 2. Traccia le semirette che congiungono i punti di controllo 1 -> 2 -> 3. Nella dimostrazione vista sopra questi sono <span style="color:#825E28">marrone</span>.

Reviewer: ```suggestion
2. Traccia le semirette che congiungono i punti di controllo 1 -> 2 -> 3. Nella dimostrazione vista sopra questi sono di colore <span style="color:#825E28">marrone</span>.
```

3. Il parametro `t` passa da `0` a `1`. Nell'esempio sopra questo cresce con passi di `0.05`: il ciclo esegue `0, 0.05, 0.1, 0.15, ... 0.95, 1`.

    Per ognuno dei valori assunti da `t`:

    - Per ogni segmento <span style="color:#825E28">marrone</span> prendiamo un punto localizzato in maniera proporzionale fra `t` ed il suo inizio. Poiché ci sono due segmenti, avremo due punti.

        Ad esempio, per `t=0`, entrambi i punti si troverano all'inizio del segmento, per `t=0.25`, si troveranno al 25% della lunghezza del segmento, per 50% (al centro), per `t=1`, alla fine del segmento.

>     - Collega i punti. Nella figura sotto, i segmenti collegati sono colarti di <span style="color:#167490">blu</span>.

Reviewer: ```suggestion
    - Collega i punti. Nella figura sotto, i segmenti collegati sono colorati di <span style="color:#167490">blu</span>.
```

5. Poiché `t` viene valutato da `0` fino a `1`, ogni valore che assume, aggiunge un punto alla curva. L'insieme di questi punti definisce la curva di Bezier. Nella figura è la parabola rossa.

Quello appena descritto, è il processo che si applica per 3 punti. Ma il procedimento è uguale anche con 4 punti.

Qui vediamo la dimostrazione con 4 punti (i punti possono essere mossi utilizzando il mouse):

[iframe src="demo.svg?p=0,0,0.5,0,0.5,1,1,1&animate=1" height=370]

L'algoritmo applicato per 4 punti:

- Connette i punti di controllo con i segmenti: 1 -> 2, 2 -> 3, 3 -> 4. Ci saranno 3 segmenti <span style="color:#825E28">marrone</span>.
- Per ogni valore di `t` nell'intervallo da `0` a `1`:
    - Tracciamo i punti nel segmento ad una distanza `t` partendo dall'inizio. Questi punti vengono connessi, in questo modo avremo due <span style="color:#0A0">segmenti verdi</span>.
    - Su questi segmenti, tracciamo i punti in base al valore di `t`. Avremo cosi un <span style="color:#167490">segmento blu</span>.
    - Nel segmento blu, tracciamo un punto in base al valore di `t`. Nell'esempio sopra è <span style="color:red">rosso</span>.
- L'insieme di questi punti forma la curva.

> L'algoritmo è ricorsivo è ricorsivo, e può essere quindi generalizzato per un qualsiasi numero di punti di controllo.

Reviewer: ```suggestion
L'algoritmo è ricorsivo, e può essere quindi generalizzato per un qualsiasi numero di punti di controllo.
```
